qutip/core/environment.py

```diff
@@ -0,0 +1,2015 @@
+            A dictionary containing the following information about the fit.
+
+            "N"
+                The number of underdamped terms used in the fit.
+            "Nk"
+                The number of Matsubara modes included per underdamped term.
+            "fit_time"
+                The time the fit took in seconds.
+            "rsme"
+                Normalized mean squared error obtained in the fit.
+            "params"
+                The fitted parameters (array of shape Nx3).
+            "summary"
+                A string that summarizes the information about the fit.
+        """
+
+        # Process arguments
+        if tag is None and self.tag is not None:
+            tag = (self.tag, "SD Fit")
+
+        if target_rsme is None:
+            target_rsme = 0
+            Nmin = Nmax
+        else:
+            Nmin = 1
+
+        jlist = self.spectral_density(wlist)
+        if guess is None and lower is None and upper is None:
+            guess, lower, upper = _default_guess_sd(wlist, jlist)
+
+        # Fit
+        start = time()
+        rmse, params = iterated_fit(
+            _sd_fit_model, 3, wlist, jlist, target_rsme, guess,
+            Nmin, Nmax, lower, upper
+        )
+        end = time()
+        fit_time = end - start
+
+        # Generate summary
+        N = len(params)
+        summary = _fit_summary(
+            fit_time, rmse, N, "the spectral density", params
+        )
+        fit_info = {
+            "N": N, "Nk": Nk, "fit_time": fit_time, "rmse": rmse,
+            "params": params, "summary": summary}
+
+        ckAR, vkAR, ckAI, vkAI = [], [], [], []
+        # Finally, generate environment and return
+        for a, b, c in params:
+            lam = np.sqrt(a + 0j)
+            gamma = 2 * b
+            w0 = np.sqrt(c**2 + b**2)
+
+            env = UnderDampedEnvironment(self.T, lam, gamma, w0)
+            coeffs = env._matsubara_params(Nk)
+            ckAR.extend(coeffs[0])
+            vkAR.extend(coeffs[1])
+            ckAI.extend(coeffs[2])
+            vkAI.extend(coeffs[3])
+
+        approx_env = ExponentialBosonicEnvironment(
+            ckAR, vkAR, ckAI, vkAI, T=self.T, tag=tag)
+        return approx_env, fit_info
+
+
+class _BosonicEnvironment_fromCF(BosonicEnvironment):
+    def __init__(self, C, tlist, tMax, T, tag):
+        super().__init__(T, tag)
+        self._cf = _complex_interpolation(C, tlist, 'correlation function')
+        if tlist is not None:
+            self.tMax = max(np.abs(tlist[0]), np.abs(tlist[-1]))
+        else:
+            self.tMax = tMax
+
+    def correlation_function(self, t, **kwargs):
+        t = np.array(t, dtype=float)
+        result = np.zeros_like(t, dtype=complex)
+        positive_mask = (t >= 0)
+        non_positive_mask = np.invert(positive_mask)
+
+        result[positive_mask] = self._cf(t[positive_mask])
+        result[non_positive_mask] = np.conj(
+            self._cf(-t[non_positive_mask])
```

Comment:
just a reminder, this still breaks from_correlation_function() baths defined only for positive times

Reply:
Hey, 

Maybe we could discuss this point! I think the issue you were facing was related to the fitting when the function you pass is all zeros and not this bit.  I sent a notebook with that example working after the previous commit through email, it's the email where I mentioned I had a problem with Bloch-Redfield in the examples and that I forgot to add combine to functions in this commit

Both Paul's previous notebook and the tests here are done for positive times only! I did say that may be the issue when you mentioned it on line, but I should have thought a bit more before replying

If this is still an issue let me know!